steering_controllers/steering_controllers/src/steering_controllers.cpp

```diff
@@ -0,0 +1,515 @@
+#include "tf2/transform_datatypes.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+
+namespace
+{  // utility
+
+// TODO(destogl): remove this when merged upstream
+// Changed services history QoS to keep all so we don't lose any client service calls
+static constexpr rmw_qos_profile_t rmw_qos_profile_services_hist_keep_all = {
+  RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+  1,  // message queue depth
+  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
+  RMW_QOS_POLICY_DURABILITY_VOLATILE,
+  RMW_QOS_DEADLINE_DEFAULT,
+  RMW_QOS_LIFESPAN_DEFAULT,
+  RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
+  RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
+  false};
+
+using ControllerReferenceMsg = steering_controllers::SteeringControllers::ControllerReferenceMsg;
+
+// called from RT control loop
+void reset_controller_reference_msg(
+  const std::shared_ptr<ControllerReferenceMsg> & msg,
+  const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> & node)
+{
+  msg->header.stamp = node->now();
+  msg->twist.linear.x = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.linear.y = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.linear.z = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.x = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.y = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.z = std::numeric_limits<double>::quiet_NaN();
+}
+
+}  // namespace
+
+namespace steering_controllers
+{
+SteeringControllers::SteeringControllers() : controller_interface::ChainableControllerInterface() {}
+
+controller_interface::CallbackReturn SteeringControllers::on_init()
+{
+  try
+  {
+    param_listener_ = std::make_shared<steering_controllers::ParamListener>(get_node());
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during controller's init with message: %s \n", e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn SteeringControllers::set_interface_numbers(
+  size_t nr_state_itfs = 2, size_t nr_cmd_itfs = 2, size_t nr_ref_itfs = 2)
+{
+  nr_state_itfs_ = nr_state_itfs;
+  nr_cmd_itfs_ = nr_cmd_itfs;
+  nr_ref_itfs_ = nr_ref_itfs;
+}
+
+controller_interface::CallbackReturn SteeringControllers::configure_odometry()
+{
+  params_ = param_listener_->get_params();
+
+  const double wheel_seperation = params_.wheel_separation_multiplier * params_.wheel_separation;
+  const double wheel_radius = params_.wheel_radius_multiplier * params_.wheel_radius;
+  const double wheelbase = params_.wheelbase_multiplier * params_.wheelbase;
+  odometry_.set_wheel_params(wheel_seperation, wheel_radius, wheelbase);
+  odometry_.set_velocity_rolling_window_size(params_.velocity_rolling_window_size);
+
+  set_interface_numbers();
+
+  RCLCPP_INFO(get_node()->get_logger(), "odom configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
```

Comment:
I propose to have this method as pure virtual and this has to be implemented by the Bicycle, Tricycle and Ackerman controller depending on a specific kinematics.

Reply:
Makes sense, this way there is no "default" controller.